### PR TITLE
Refactor adding 'lang' query parameter - IDVA3-1281

### PR DESCRIPTION
**Pull request by hepsimo**
- fix Back link on PSC type screen to correct URI (including companyNumber)
- add queryParams.ts module to support adding multiple params generically
- replace URI string manipulation using queryParams.ts
- change tsconfig.json to help write unit tests:
  1. add compiler checking of /test folder tree (enables auto import suggestions)
  2. add jest to types (fixed compiler not finding jest identifiers, e.g. "expect", after step 1 change)
  3. add 'build' version of tsconfig to exclude test sources from dist/ folder

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by PhilipGlover** [comment]
I can't get the service to start:

**Comment by PhilipGlover** [comment]
The service won't start for me:
```
2024-04-08 09:05:02 Error: Cannot find module '/app/dist/server.js'
2024-04-08 09:05:02     at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
2024-04-08 09:05:02     at Module._load (node:internal/modules/cjs/loader:985:27)
2024-04-08 09:05:02     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
2024-04-08 09:05:02     at node:internal/main/run_main_module:28:49 {
2024-04-08 09:05:02   code: 'MODULE_NOT_FOUND',
2024-04-08 09:05:02   requireStack: []
2024-04-08 09:05:02 }
2024-04-08 09:05:02 
2024-04-08 09:05:02 Node.js v20.11.0
2024-04-08 09:05:02 [nodemon] app crashed - waiting for file changes before starting...
```

File: tsconfig.json [hunk 1]
```diff
@@ -27,7 +29,8 @@
   },
   "lib": ["esnext"],
   "include": [
-    "./src/**/*.ts"
+    "./src/**/*.ts",
+    "./test/**/*.ts"
```
Comment (PhilipGlover) [inline tsconfig.json:33]: this addition, as we saw, is causing the `dist/` folder to now contain `/src` and `/test` sub folders.  Which means the service won't currently start.

Comment (hepsimo) [reply tsconfig.json:33]: Thanks, i'll fix this. There is a suitable solution here: https://stackoverflow.com/a/60223923

File: src/routers/handlers/confirmCompany/confirmCompany.ts [hunk 2]
```diff
@@ -29,12 +30,13 @@ export class ConfirmCompanyHandler extends GenericHandler<ConfirmCompanyViewData
...
         const address = buildAddress(companyProfile);
+        const currentUrl = addSearchParams(PrefixedUrls.CONFIRM_COMPANY, { companyNumber: companyProfile.companyNumber, lang });
 
         return {
             ...baseViewData,
             ...getLocaleInfo(locales, lang),
-            currentUrl: PrefixedUrls.CONFIRM_COMPANY + "?lang=" + lang,
-            backURL: PrefixedUrls.START + "?lang=" + lang,
+            currentUrl,
+            backURL: addSearchParams(PrefixedUrls.START, { lang }),
```
[the remark this answers is not in the document]
Comment (hepsimo) [reply src/routers/handlers/confirmCompany/confirmCompany.ts:39]: Agreed, i'll add this.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: package.json [hunk 3]
```diff
@@ -10,8 +10,8 @@
     "scripts": {
       "prebuild": "rm -rf ./dist",
       "build:assets": "gulp build",
-      "build:server": "tsc && cp -r src/views dist/ && cp -r src/config dist/",
-      "build": "tsc && cp -r src/views dist/",
+      "build:server": "tsc -p tsconfig-build.json && cp -r src/views dist/ && cp -r src/config dist/",
+      "build": "tsc -p tsconfig-build.json && cp -r src/views dist/",
```
Comment (PhilipGlover) [inline package.json:14]: as discussed maybe `tsc --project` clearer.. I had to look up what `-p` was short for.

**Review by PhilipGlover — APPROVED** [review]
- have successfully built and run
- back link navigation from confirm company screen now works, and changed to go to Company lookup
- welsh navigation works
- new tests runs and coverage 100%
Screenshot 2024-04-08 at 3 54 39 PM
- approving though see comment on `tsc` command in package script

File: package.json (hunk 3 above)
Comment (hepsimo) [reply package.json:14]: Agreed, i'll change this.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Pull request merged**